bootstrap/bootstrap.go

```diff
@@ -319,132 +335,296 @@ func (b *Bootstrap) StartAPIServer(ctx context.Context) error {
 		pullAPI,
 		debugAPI,
 		walletAPI,
-		b.config,
+		cfg,
 	)
 
-	if err := b.server.EnableRPC(supportedAPIs); err != nil {
-		return err
+	if err := server.EnableRPC(supportedAPIs); err != nil {
+		return nil, err
 	}
 
-	if b.config.WSEnabled {
-		if err := b.server.EnableWS(supportedAPIs); err != nil {
-			return err
+	if cfg.WSEnabled {
+		if err := server.EnableWS(supportedAPIs); err != nil {
+			return nil, err
 		}
 	}
 
-	if err := b.server.SetListenAddr(b.config.RPCHost, b.config.RPCPort); err != nil {
-		return err
-	}
-
-	if err := b.server.Start(); err != nil {
-		return err
+	if err := server.SetListenAddr(cfg.RPCHost, cfg.RPCPort); err != nil {
+		return nil, err
 	}
 
-	b.logger.Info().Msgf("API server started: %s", b.server.ListenAddr())
-	return nil
+	return server, nil
 }
 
-func (b *Bootstrap) StopAPIServer() {
-	if b.server == nil {
-		return
+func (fnb *EVMGatewayNodeBuilder) eventIngestionEngineComponent(cfg config.Config) (module.ReadyDoneAware, error) {
+	l := fnb.Logger.With().Str("component", "bootstrap-ingestion").Logger()
+	l.Info().Msg("bootstrap starting event ingestion")
+
+	// get latest cadence block from the network and the database
+	chainLatestBlock, err := fnb.Client.GetLatestBlock(context.Background(), true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest cadence block: %w", err)
 	}
-	b.logger.Warn().Msg("shutting down API server")
-	b.server.Stop()
-}
 
-func (b *Bootstrap) StartMetricsServer(ctx context.Context) error {
-	b.logger.Info().Msg("bootstrap starting metrics server")
+	gatewayLatestHeight, err := fnb.Storages.Blocks.LatestCadenceHeight()
+	if err != nil {
+		return nil, err
+	}
 
-	b.metrics = flowMetrics.NewServer(b.logger, uint(b.config.MetricsPort))
-	err := util.WaitClosed(ctx, b.metrics.Ready())
+	// make sure the provided block to start the indexing can be loaded
+	_, err = fnb.Client.GetBlockHeaderByHeight(context.Background(), chainLatestBlock.Height)
 	if err != nil {
-		return fmt.Errorf("failed to start metrics server: %w", err)
+		return nil, fmt.Errorf(
+			"failed to get provided cadence height %d: %w",
+			chainLatestBlock.Height,
+			err,
+		)
 	}
 
-	return nil
-}
+	l.Info().
+		Uint64("chain-cadence-height", chainLatestBlock.Height).
+		Uint64("gateway-cadence-height", gatewayLatestHeight).
+		Uint64("missed-heights", chainLatestBlock.Height-gatewayLatestHeight).
```

Comment:
Simply avoid underflow
```suggestion
		Int64("missed-heights", chainLatestBlock.Height-gatewayLatestHeight).
```